docs: Add more documentation (#18)
aneoconsulting/ArmoniK.Extensions.Cpp: .docs/content/guide/3.end2end.md

<!-- @case-police-ignore Api -->
<!-- @case-police-ignore Sdk -->

# Build and configure your own client/worker test

You can also develop your own client to submit jobs to ArmoniK and invoke desired services  which are embedded in the worker-shared library.

## Client test

The SessionService class in ```ArmoniK::Sdk::Client``` namespace is used to create a client. This class essentially contains methods for sending tasks to ArmoniK and retrieving the results. These methods are ```Submit``` and ```WaitResults``` respectively.

### Configure client

> The constructor of the SessionService class takes as arguments a properties and a logger objects. The properties object as defined in ```ArmoniK::Sdk::Common``` namespace contain the configuration and a session task options objects.

Reviewer: a property or properties?

Author: I wanted to keep the name of the variable in the code, so its properties. But I will highlight it to avoid confusion.

### Worker service implementation

To create an ArmoniK Worker Service, a shared library exposing 5 mandatory symbols must be created (cf ArmoniK SDK Interface). This library must then be deployed either by copying it to the shared folder ARMONIK_SHARED_HOST_PATH, or by copying it in the Docker container run by the compute plane. The latter option may require overriding the default application search path by giving the environment variable Worker__ApplicationBasePath to the directory where the library is located.

The application_name and application_version task options are then determined by the name of the library such that “application_name.application_version” or “application_name” is the name of the library file. This library will be loaded and unloaded according to the task options of the task being processed by the worker. The different calls to the symbols and the expected lifetimes of the service and sessions within a worker pod are described in the flowchart presented in the appendix section.

### Setup worker shared lib

> It is possible set up a shared library in order to use the ArmoniK Worker. To aid in the creation of this library, it can link the ArmoniK.SDK.Worker static library and expose its symbols. The ArmoniK.SDK.Worker shared library is to be used with C++ workers (C++14 or later).

Reviewer 2: ```suggestion
It is possible to set up a shared library in order to extend the ArmoniK Worker. In order to facilitate writing this shared library it is recommended to link your own code to the ArmoniK.SDK.Worker static library. It provides default implementation of mandatory symbols (```armonik_enter_session```, ```armonik_enter_session```,  ```armonik_call```, ```armonik_leave_session``` and ```armonik_destroy_service```) . These symbols will be explained later in this section.  Please note that ArmoniK.SDK.Worker shared library is to be used with C++ workers (C++14 or later).
```

> To do so, an implementation of a service class inheriting of ```ArmoniK::Sdk::Worker::ServiceBase``` must be provided. An example of this kind of implementation can be found in [Echo service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/EchoService.h) or [Addition service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/AdditionService.h). The examples provide the service class implementing the methods called when entering, calling or leaving a session using the ```ArmoniK::Sdk::Worker```. When using this approach, users may only provide the implementation of the creation of ```armonik_create_service``` as in [here](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/src/ServiceDispatch.cpp)

Reviewer 2: ```suggestion
To do so, the developer can provide an implementation of a service class inheriting of ```ArmoniK::Sdk::Worker::ServiceBase``` . An example of this kind of implementation can be found in [Echo service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/EchoService.h) or [Addition service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/AdditionService.h). The examples provide the service class implementing the methods called when entering, calling or leaving a session using the ```ArmoniK::Sdk::Worker```. When using this approach, users may only provide the implementation of the creation of ```armonik_create_service``` as in [here](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/src/ServiceDispatch.cpp)
```

> To do so, an implementation of a service class inheriting of ```ArmoniK::Sdk::Worker::ServiceBase``` must be provided. An example of this kind of implementation can be found in [Echo service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/EchoService.h) or [Addition service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/AdditionService.h). The examples provide the service class implementing the methods called when entering, calling or leaving a session using the ```ArmoniK::Sdk::Worker```. When using this approach, users may only provide the implementation of the creation of ```armonik_create_service``` as in [here](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/src/ServiceDispatch.cpp)

Reviewer 2: ```suggestion
To do so, an implementation of a service class inheriting of ```ArmoniK::Sdk::Worker::ServiceBase``` must be provided. The developer could override previously cited symbols or keep their default implementations as they are. However ```armonik_create_service``` symbol is not provided by ```ServiceBase``` and must be implemented by the user.  An example of this kind of implementation can be found in [Echo service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/EchoService.h) or [Addition service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/AdditionService.h). The examples provide the service class implementing the methods called when entering, calling or leaving a session using the ```ArmoniK::Sdk::Worker```. When using this approach, users may only provide the implementation of the creation of ```armonik_create_service``` as in [here](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/src/ServiceDispatch.cpp)
```

> It is also possible to set up the shared library without inheriting ArmoniK::Sdk::Worker::ServiceBase. In that scenario, the implementation of the following mandatory methods must be provided:

Reviewer 2: ```suggestion
Another way to set up the shared library without inheriting ArmoniK::Sdk::Worker::ServiceBase consists in providing an implementation of the following mandatory methods:
```

```armonik_create_service``` This symbol is mandatory and allows to create a service based on its name and namespace. This function should be implemented by the library and return a service context. This service context will be passed to the subsequent calls to other symbols. The returned context may be null if desired.
```armonik_enter_session``` This symbol will be called when entering a session. This function returns a user defined session context.
```armonik_call``` This symbol will be called to request the execution of a method. This function returns the status of the ArmoniK call. The function takes multiple arguments including a Callback provided by ArmoniK to send the result of the execution or details on a failure. See the symbols documentation for more details [here](https://aneoconsulting.github.io/ArmoniK.Extensions.Cpp/doxygen/html/ArmoniKSDKInterface_8h.html).
```armonik_leave_session``` This function is called when leaving a session. It should free the resources of the session context.
```armonik_destroy_service``` This function is called when the service should be destroyed. It should free the resources of the service context.

> If the user is using the ```ArmoniK::Sdk::Worker``` library, all these symbols have a default implementation except ```armonik_create_service```, which must be implemented by the user. The default implementations can also be overridden to add functionality. If ```ArmoniK::Sdk::Worker``` is not used, then the user provides the implementation of all symbols.

Reviewer 2: ```suggestion
```
